src/coreclr/src/interop/comwrappers.cpp

```diff
@@ -0,0 +1,662 @@
+{
+    _ASSERTE(cxtMaybe != nullptr);
+
+    // Convert the supplied context
+    char* cxtRaw = reinterpret_cast<char*>(cxtMaybe);
+    cxtRaw -= sizeof(NativeObjectWrapperContext);
+    NativeObjectWrapperContext* cxt = reinterpret_cast<NativeObjectWrapperContext*>(cxtRaw);
+
+#ifdef _DEBUG
+    _ASSERTE(cxt->_sentinal == LiveContextSentinal);
+#endif
+
+    return cxt;
+}
+
+HRESULT NativeObjectWrapperContext::Create(
+    _In_ IUnknown* external,
+    _In_ CreateObjectFlags flags,
+    _In_ size_t runtimeContextSize,
+    _Outptr_ NativeObjectWrapperContext** context)
+{
+    _ASSERTE(external != nullptr && context != nullptr);
+
+    HRESULT hr;
+
+    ComHolder<IReferenceTracker> trackerObject;
+    if ((flags & CreateObjectFlags::TrackerObject) == CreateObjectFlags::TrackerObject)
+    {
+        hr = external->QueryInterface(&trackerObject);
+        if (SUCCEEDED(hr))
+            RETURN_IF_FAILED(TrackerObjectManager::OnIReferenceTrackerFound(trackerObject));
+    }
+
+    // Allocate memory for the RCW
+    char* cxtMem = (char*)InteropLibImports::MemAlloc(sizeof(NativeObjectWrapperContext) + runtimeContextSize, AllocScenario::NativeObjectWrapper);
+    if (cxtMem == nullptr)
+        return E_OUTOFMEMORY;
+
+    void* runtimeContext = cxtMem + sizeof(NativeObjectWrapperContext);
+
+    // Contract specifically requires zeroing out runtime context.
+    std::memset(runtimeContext, 0, runtimeContextSize);
+
+    NativeObjectWrapperContext* contextLocal = new (cxtMem) NativeObjectWrapperContext{ runtimeContext, trackerObject };
+
+    if (trackerObject != nullptr)
+    {
+        // Inform the tracker object manager
+        _ASSERTE((flags & CreateObjectFlags::TrackerObject) == CreateObjectFlags::TrackerObject);
+        hr = TrackerObjectManager::AfterWrapperCreated(trackerObject);
+        if (FAILED(hr))
+        {
+            Destroy(contextLocal);
+            return hr;
+        }
+    }
+
+    *context = contextLocal;
+    return S_OK;
+}
+
+void NativeObjectWrapperContext::Destroy(_In_ NativeObjectWrapperContext* wrapper)
+{
+    _ASSERTE(wrapper != nullptr);
+
+    // Manually trigger the destructor since placement
+    // new was used to allocate the object.
+    wrapper->~NativeObjectWrapperContext();
+    InteropLibImports::MemFree(wrapper, AllocScenario::NativeObjectWrapper);
+}
+
+NativeObjectWrapperContext::NativeObjectWrapperContext(_In_ void* runtimeContext, _In_opt_ IReferenceTracker* trackerObject)
+    : _trackerObject{ trackerObject }
+    , _runtimeContext{ runtimeContext }
+    , _isValidTracker{ (trackerObject != nullptr ? TRUE : FALSE) }
+#ifdef _DEBUG
+    , _sentinal{ LiveContextSentinal }
+#endif
+{
+    if (_isValidTracker == TRUE)
+        (void)_trackerObject->AddRef();
+}
+
+NativeObjectWrapperContext::~NativeObjectWrapperContext()
+{
+    DisconnectTracker();
+
+#ifdef _DEBUG
+    _sentinal = DeadContextSentinal;
```

Comment:
Nit: Should all of the `sentin*a*l`  be spelled `sentin*e*l` in this file?

Reply:
Sigh... Thanks @MichalStrehovsky!